// ngx-highlightjs/esm2015/lib/highlight-children.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Directive, Input, Renderer2, ElementRef, NgZone } from '@angular/core';
import { HighlightJS } from './highlight.service';
export class HighlightChildren {
    /**
     * @param {?} _zone
     * @param {?} _el
     * @param {?} _hljs
     * @param {?} _renderer
     */
    constructor(_zone, _el, _hljs, _renderer) {
        this._zone = _zone;
        this._el = _el;
        this._hljs = _hljs;
        this._renderer = _renderer;
        this._observers = [];
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.highlightChildren(this.selector);
    }
    /**
     * Highlight a code block
     * @param {?} el Code block element
     * @return {?}
     */
    highlightElement(el) {
        // Remove element classes
        el.className = '';
        this._renderer.addClass(el, 'hljs');
        this._hljs.highlightBlock(el);
    }
    /**
     * Highlight multiple code blocks
     * @param {?} selector elements selector
     * @return {?}
     */
    highlightChildren(selector) {
        this._zone.runOutsideAngular(() => {
            /** @type {?} */
            const elementsToHighlight = this._el.nativeElement.querySelectorAll(selector || 'code');
            elementsToHighlight.forEach((element) => {
                // Highlight element when text is present
                /** @type {?} */
                const observer = new MutationObserver(() => {
                    if (element.childNodes.length === 1 &&
                        element.childNodes[0].nodeName === '#text') {
                        this.highlightElement(element);
                    }
                });
                observer.observe(element, { childList: true });
                this._observers.push(observer);
            });
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._observers.forEach((observer) => observer.disconnect());
        this._observers = null;
    }
}
HighlightChildren.decorators = [
    { type: Directive, args: [{
                selector: '[highlightChildren]'
            },] }
];
/** @nocollapse */
HighlightChildren.ctorParameters = () => [
    { type: NgZone },
    { type: ElementRef },
    { type: HighlightJS },
    { type: Renderer2 }
];
HighlightChildren.propDecorators = {
    selector: [{ type: Input, args: ['highlightChildren',] }]
};
if (false) {
    /** @type {?} */
    HighlightChildren.prototype._observers;
    /** @type {?} */
    HighlightChildren.prototype.selector;
    /** @type {?} */
    HighlightChildren.prototype._zone;
    /** @type {?} */
    HighlightChildren.prototype._el;
    /** @type {?} */
    HighlightChildren.prototype._hljs;
    /** @type {?} */
    HighlightChildren.prototype._renderer;
}
