// ngx-highlightjs/esm5/lib/highlight-children.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Directive, Input, Renderer2, ElementRef, NgZone } from '@angular/core';
import { HighlightJS } from './highlight.service';
var HighlightChildren = /** @class */ (function () {
    function HighlightChildren(_zone, _el, _hljs, _renderer) {
        this._zone = _zone;
        this._el = _el;
        this._hljs = _hljs;
        this._renderer = _renderer;
        this._observers = [];
    }
    /**
     * @return {?}
     */
    HighlightChildren.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.highlightChildren(this.selector);
    };
    /**
     * Highlight a code block
     * @param el Code block element
     */
    /**
     * Highlight a code block
     * @param {?} el Code block element
     * @return {?}
     */
    HighlightChildren.prototype.highlightElement = /**
     * Highlight a code block
     * @param {?} el Code block element
     * @return {?}
     */
    function (el) {
        // Remove element classes
        el.className = '';
        this._renderer.addClass(el, 'hljs');
        this._hljs.highlightBlock(el);
    };
    /**
     * Highlight multiple code blocks
     * @param selector elements selector
     */
    /**
     * Highlight multiple code blocks
     * @param {?} selector elements selector
     * @return {?}
     */
    HighlightChildren.prototype.highlightChildren = /**
     * Highlight multiple code blocks
     * @param {?} selector elements selector
     * @return {?}
     */
    function (selector) {
        var _this = this;
        this._zone.runOutsideAngular(function () {
            /** @type {?} */
            var elementsToHighlight = _this._el.nativeElement.querySelectorAll(selector || 'code');
            elementsToHighlight.forEach(function (element) {
                // Highlight element when text is present
                /** @type {?} */
                var observer = new MutationObserver(function () {
                    if (element.childNodes.length === 1 &&
                        element.childNodes[0].nodeName === '#text') {
                        _this.highlightElement(element);
                    }
                });
                observer.observe(element, { childList: true });
                _this._observers.push(observer);
            });
        });
    };
    /**
     * @return {?}
     */
    HighlightChildren.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._observers.forEach(function (observer) { return observer.disconnect(); });
        this._observers = null;
    };
    HighlightChildren.decorators = [
        { type: Directive, args: [{
                    selector: '[highlightChildren]'
                },] }
    ];
    /** @nocollapse */
    HighlightChildren.ctorParameters = function () { return [
        { type: NgZone },
        { type: ElementRef },
        { type: HighlightJS },
        { type: Renderer2 }
    ]; };
    HighlightChildren.propDecorators = {
        selector: [{ type: Input, args: ['highlightChildren',] }]
    };
    return HighlightChildren;
}());
export { HighlightChildren };
if (false) {
    /** @type {?} */
    HighlightChildren.prototype._observers;
    /** @type {?} */
    HighlightChildren.prototype.selector;
    /** @type {?} */
    HighlightChildren.prototype._zone;
    /** @type {?} */
    HighlightChildren.prototype._el;
    /** @type {?} */
    HighlightChildren.prototype._hljs;
    /** @type {?} */
    HighlightChildren.prototype._renderer;
}
